lib/mail_retriever.rb

```diff
@@ -0,0 +1,73 @@
+require 'mail'
+
+# USAGE:
+#
+# manager = MailRetriever.new(
+#   config: {
+#     user_name: 'reporting@pensionwise.gov.uk',
+#     password: <password>
+#   }
+# )
+#
+# PASSWORD:
+#
+# Generate an app specific password following insteructions:
+#   https://support.google.com/mail/answer/185833?hl=en&rd=1
+#
+# manager.all(
+#   search_keys: [
+#     'SUBJECT', 'TP Daily Call Data',
+#     'FROM', '@pensionwise.gov.uk',
+#   ],
+#   file_name_regexp: /DWP Daily Data File.*\.xlsx/
+# )
+#
+
+class MailRetriever
+  DEFAULT_CONFIG = {
+    address: 'imap.googlemail.com',
+    port: 993,
+    enable_ssl: true,
+    user_name: nil,
+    password: nil
+  }.freeze
+
+  def initialize(config:)
+    Mail.defaults do
+      retriever_method :imap, DEFAULT_CONFIG.merge(config)
+    end
+  end
+
+  def most_recent(search_keys: ['All'], file_name_regexp:)
+    search(search_keys: search_keys, file_name_regexp: file_name_regexp, count: 1).first
+  end
+
+  def search(search_keys: ['All'], file_name_regexp:, count: :all)
+    results = []
+
+    Mail.last(count: count, order: :desc, keys: search_keys) do |mail, _imap, uid|
+      attachment = mail.attachments.detect { |a| a.filename =~ file_name_regexp }
+
+      results << EmailAttachment.new(uid: uid, attachment: attachment, subject: mail.subject) if attachment
```

Comment:
Can we promote the postfix modifier to a regular `if` block? It's kinda lost way out there.